pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -134,151 +143,123 @@ lazy_static! {
     .expect("failed to register pageserver image sync time histogram vec");
 }
 
+static SYNC_QUEUE: OnceCell<SyncQueue> = OnceCell::new();
+
 /// Wraps mpsc channel bits around into a queue interface.
 /// mpsc approach was picked to allow blocking the sync loop if no tasks are present, to avoid meaningless spinning.
-mod sync_queue {
-    use std::{
-        collections::{hash_map, HashMap, HashSet},
-        num::NonZeroUsize,
-        ops::ControlFlow,
-        sync::atomic::{AtomicUsize, Ordering},
-    };
-
-    use anyhow::anyhow;
-    use once_cell::sync::OnceCell;
-    use tokio::sync::mpsc::{error::TryRecvError, UnboundedReceiver, UnboundedSender};
-    use tracing::{debug, warn};
-
-    use super::SyncTask;
-    use utils::zid::ZTenantTimelineId;
-
-    static SENDER: OnceCell<UnboundedSender<(ZTenantTimelineId, SyncTask)>> = OnceCell::new();
-    static LENGTH: AtomicUsize = AtomicUsize::new(0);
-
-    /// Initializes the queue with the given sender channel that is used to put the tasks into later.
-    /// Errors if called more than once.
-    pub fn init(sender: UnboundedSender<(ZTenantTimelineId, SyncTask)>) -> anyhow::Result<()> {
-        SENDER
-            .set(sender)
-            .map_err(|_sender| anyhow!("sync queue was already initialized"))?;
-        Ok(())
-    }
+struct SyncQueue {
+    len: AtomicUsize,
+    max_timelines_per_batch: NonZeroUsize,
+    sender: UnboundedSender<(ZTenantTimelineId, SyncTask)>,
+    receiver: Mutex<UnboundedReceiver<(ZTenantTimelineId, SyncTask)>>,
+}
 
-    /// Adds a new task to the queue, if the queue was initialized, returning `true` on success.
-    /// On any error, or if the queue was not initialized, the task gets dropped (not scheduled) and `false` is returned.
-    pub fn push(sync_id: ZTenantTimelineId, new_task: SyncTask) -> bool {
-        if let Some(sender) = SENDER.get() {
-            match sender.send((sync_id, new_task)) {
-                Err(e) => {
-                    warn!("Failed to enqueue a sync task: the receiver is dropped: {e}");
-                    false
-                }
-                Ok(()) => {
-                    LENGTH.fetch_add(1, Ordering::Relaxed);
-                    true
-                }
-            }
-        } else {
-            warn!("Failed to enqueue a sync task: the sender is not initialized");
-            false
+impl SyncQueue {
+    fn new(max_timelines_per_batch: NonZeroUsize) -> Self {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        Self {
+            len: AtomicUsize::new(0),
+            max_timelines_per_batch,
+            sender,
+            receiver: Mutex::new(receiver),
```

Comment:
This is an unfortunate combination of sync and async worlds.
The root cause is the fact that all receive methods on the `receiver` are `&mut self`, such as `try_recv(&mut self) -> ...` and the fact that we have to keep a set of static pub methods for submitting sync messages (upload/download/delete), ideally without blocking too much. 
Those static methods have to be synchronous, since are called from sync parts of pageserver.

For those static methods, a queue has to be placed into a shared state, we use `lazy_static` or `once_cell` and `Mutex/RwLock/RefCell` for interior mutability: `once_cell` allows to retrieve the data as `fn get(&self) -> Option<&T>`, lazy static also returns `&` and we need `&mut` on the channel.

But either those mutexes have to come from `std::sync` module or I have to add some static runtime to `.await` on the tokio ones.

----

Or I could push the async mutex down where it is now and call it in the `async` code during the task processing.

Any other, better ways? Maybe I should use a static runtime after all?

Reply:
> we have to keep a set of static pub methods for submitting sync messages

There is no need to access the receiver during submission, right? Are there any problems with decoupling the receiver from this struct entirely? Looks like we can return `(Self, Receiver)` from `SyncQueue::new` and then move the receiver to the sync thread without any synchronization requierd in the future

Reply:
That actually worked not that bad, even though a bit verbose, thanks.